#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [1]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [26]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("_Train__202304301950.csv")


## Data exploration

In [27]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
 13  column1           0 non-null      float64
dtypes: float64(3), int64(1), object(10)
memory usage: 99.9+ KB


In [28]:
train.head()

laptop_ID Company            Product   TypeName  Inches  \
0       1104    Acer     Aspire ES1-523   Notebook    15.6   
1        114    Dell             XPS 13  Ultrabook    13.3   
2        172    Acer    Aspire A517-51G   Notebook    17.3   
3        918      HP      Elitebook 820    Netbook    12.5   
4        447  Lenovo  IdeaPad 320-15AST   Notebook    17.3   

                   ScreenResolution                         Cpu  Ram  \
0                          1366x768   AMD A8-Series 7410 2.2GHz  4GB   
1  Quad HD+ / Touchscreen 3200x1800  Intel Core i7 7560U 2.4GHz  8GB   
2       IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz  8GB   
3                 Full HD 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   
4                          1600x900   AMD A6-Series 9220 2.5GHz  8GB   

      Memory                           Gpu       OpSys  Weight  Price_euros  \
0  500GB HDD                 AMD Radeon R5  Windows 10   2.4kg        387.0   
1  256GB SSD  Intel Iris Plus Graphics 640  Windows 10  1.23kg       1379.0   
2  256GB SSD          Nvidia GeForce MX150  Windows 10     3kg        854.0   
3  512GB SSD         Intel HD Graphics 620  Windows 10  1.26kg       1483.0   
4    1TB HDD                 AMD Radeon R4  Windows 10   2.8kg        519.0   

   column1  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

## Data processing

In [29]:
train['Ram'] = train['Ram'].str.replace('GB', '').astype(float)


In [30]:
train['Weight'] = train['Weight'].str.replace('kg', '').astype(float)

In [31]:
train = train.drop("column1", axis=1)

In [32]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    float64
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    float64
 12  Price_euros       912 non-null    float64
dtypes: float64(4), int64(1), object(8)
memory usage: 92.8+ KB


-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [34]:
train = train[['laptop_ID','Inches', "Ram", "Weight",'Price_euros']]

In [35]:
train

laptop_ID  Inches   Ram  Weight  Price_euros
0         1104    15.6   4.0    2.40       387.00
1          114    13.3   8.0    1.23      1379.00
2          172    17.3   8.0    3.00       854.00
3          918    12.5   8.0    1.26      1483.00
4          447    17.3   8.0    2.80       519.00
..         ...     ...   ...     ...          ...
907        996    15.6   6.0    2.20       549.00
908        731    17.3  32.0    4.36      3659.40
909        276    17.3  16.0    3.20      1549.00
910        585    17.3  16.0    4.14      2729.00
911       1173    15.6   4.0    1.91       478.89

[912 rows x 5 columns]

In [39]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 4)

In [40]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [41]:
from sklearn.model_selection import train_test_split

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [43]:
X_train

laptop_ID  Inches   Ram  Weight
25          26    15.6   4.0    2.30
84         776    15.6  16.0    2.72
10        1263    15.6  16.0    2.50
342        182    15.6   8.0    1.90
890        321    13.3  16.0    1.29
..         ...     ...   ...     ...
106        394    15.6   8.0    2.00
270        441    15.6  16.0    2.50
860        955    17.3   4.0    2.80
435        210    15.6   6.0    2.30
102        673    15.6   4.0    2.10

[729 rows x 4 columns]

### 3. Asignar el modelo (vacío) a una variable


In [44]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

In [45]:
model.fit(X,y)

LinearRegression()

In [46]:
pred = model.predict(X_test)
pred

array([1048.17859406,  523.89326864,  960.74965996,  949.80366013,
       1910.89478536,  823.89833142,  669.00073428, 1180.49800971,
        967.19371385, 1209.93937539, 1231.29826087,  454.19165263,
        733.29081296, 1232.52907675, 1224.03054204, 1787.87649563,
       2028.35315935, 1286.03912245,  912.48873784,  432.32396243,
        876.35994306,  675.28378668,  985.2790246 , 1434.91424163,
        888.41773956,  498.73967733,  799.96247685, 1753.46427972,
        964.11177503,  660.07206812, 1167.90945144, 1120.88458875,
       1139.63522827, 1350.01744613, 1196.21753832,  541.77980071,
       1127.59462039,  810.85505919, 1183.29570288, 1134.22643231,
       1129.37716877, 1017.50862278, 1083.04367409, 1776.62649074,
       1132.049556  ,  661.6819446 , 1864.61332412,  634.09787705,
        993.64945668, 1010.60416969, 2195.97786521, 1120.46763827,
        558.70370765, 1769.83188367,  790.97631561, 1173.81040896,
       2151.15833584,  986.5985083 , 1995.65449895, 1873.71330

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [83]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE :',np.sqrt(mean_squared_error(y_test,pred)))

RMSE : 477.86077137664546


In [49]:
from sklearn import metrics

In [50]:
print('MAE:', metrics.mean_absolute_error(y_test, pred))
print('MSE:', metrics.mean_squared_error(y_test, pred))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred)))

MAE: 324.78709048522427
MSE: 228350.9168206826
RMSE: 477.86077137664546


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [61]:
X_pred = pd.read_csv("_Test__202304301950.csv")
X_pred.head()

laptop_ID   Company          Product            TypeName  Inches  \
0        750   Toshiba  Tecra X40-D-10G            Notebook    14.0   
1        633  Mediacom    SmartBook 140            Notebook    14.0   
2        363      Dell    Inspiron 7577              Gaming    15.6   
3        319      Asus    VivoBook Flip  2 in 1 Convertible    11.6   
4         14      Dell    Inspiron 3567            Notebook    15.6   

                            ScreenResolution  \
0  IPS Panel Full HD / Touchscreen 1920x1080   
1                          Full HD 1920x1080   
2                          Full HD 1920x1080   
3                       Touchscreen 1366x768   
4                          Full HD 1920x1080   

                                    Cpu  Ram              Memory  \
0            Intel Core i5 7200U 2.5GHz  8GB           256GB SSD   
1           Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
2           Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
3  Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   
4              Intel Core i3 6006U 2GHz  4GB           256GB SSD   

                       Gpu       OpSys  Weight  
0    Intel HD Graphics 620  Windows 10  1.25kg  
1        Intel HD Graphics  Windows 10   1.4kg  
2  Nvidia GeForce GTX 1050  Windows 10  2.65kg  
3    Intel HD Graphics 500  Windows 10   1.1kg  
4       AMD Radeon R5 M430  Windows 10   2.2kg

In [63]:
X_pred['Ram'] = X_pred['Ram'].str.replace('GB', '').astype(float)

In [68]:

X_pred['Weight'] = X_pred['Weight'].str.replace('kg', '').astype(float)

c:\Users\Abraham\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [69]:
X_pred.head()

laptop_ID  Inches  Ram  Weight
0        750    14.0  8.0    1.25
1        633    14.0  2.0    1.40
2        363    15.6  8.0    2.65
3        319    11.6  2.0    1.10
4         14    15.6  4.0    2.20

In [70]:
X_pred=X_pred[['laptop_ID','Inches', "Ram", "Weight"]]

In [71]:
X_pred

laptop_ID  Inches  Ram  Weight
0          750    14.0  8.0    1.25
1          633    14.0  2.0    1.40
2          363    15.6  8.0    2.65
3          319    11.6  2.0    1.10
4           14    15.6  4.0    2.20
..         ...     ...  ...     ...
386       1132    12.5  8.0    1.25
387         49    15.6  4.0    2.30
388       1243    15.6  4.0    2.65
389        924    12.5  8.0    1.09
390        726    14.0  4.0    1.40

[391 rows x 4 columns]

**¿Por qué da error?** 

In [72]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1162.51150447,  484.44460493, 1002.60464019,  578.69790028,
        518.7983766 ,  440.51878505, 1922.07857077, 1884.9836653 ,
       1093.67238005,  961.08155565, 1084.25980355, 1031.48719347,
        986.66596364,  528.09242692, 1067.35380376, 1113.62035549,
       2021.28949844,  588.84244059, 1091.45188337, 1206.74798287,
        890.37897335,  694.43459746, 1131.56532184,  641.14703176,
       1160.33170657,  921.72117601,  894.32179033,  550.73846978,
       1061.17193436,  969.92397155, 1030.94775432,  872.28344511,
       1110.06966215, 1094.12750004, 1167.11715654, 1063.86846439,
       1261.90283531, 1130.63538106, 2074.42084778, 1828.27134006,
        956.52610276,  699.64194941, 1096.12620332,  963.55839364,
       1916.01756803,  966.89610979, 1805.94786414, 1002.56572891,
       1911.50629334,  668.85311106,  556.33850934,  994.05982835,
       1123.77851326,  523.42128483,  643.31985854, 1006.82471183,
        888.61904655, 1237.95351543, 1192.30164954,  523.00102

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [43]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1097.6423004 , 1095.00755928, 1084.79793741, 1101.01825451,
        981.53500887, 1062.70472731, 1123.17875843, 1136.91660235,
       1100.68891187, 1171.05845611, 1061.49713762, 1039.67799563,
       1120.5592512 , 1084.55112248, 1067.42530516, 1144.60126397,
       1008.086748  , 1087.21311678, 1117.59516743, 1166.6672209 ,
       1136.28517025, 1137.01114934, 1201.1384173 , 1195.5395924 ,
       1155.57935199, 1095.22712104, 1205.96877603, 1141.07304191,
       1087.734768  , 1145.58929189, 1069.84048452, 1117.48538655,
       1169.08240026, 1183.02457206, 1166.99656354, 1048.59589281,
       1143.28389341, 1069.57766753, 1078.759989  , 1072.74768646,
       1196.62216731, 1063.58297435, 1030.59182845,  995.79128941,
       1065.55903019, 1129.78084514, 1071.26763597, 1167.64001493,
       1224.39673002, 1248.32896191, 1181.26807797, 1092.56512673,
       1090.69885177, 1189.17230135, 1066.32749635, 1158.8727784 ,
       1102.11606331, 1053.33170455, 1106.66997846, 1072.30856

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [73]:
sample = pd.read_csv("_Sample__202304301950.csv")

In [74]:
sample.head()

laptop_ID  Price_euros
0        750        500.0
1        726        500.0
2        633        500.0
3        363        650.0
4        319        650.0

In [75]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [76]:
submission = pd.DataFrame({"I": X_pred['laptop_ID'], "Price_euros": predictions_submit})

In [77]:
submission.head()

laptop_ID  Price_euros
0        750  1162.511504
1        726   484.444605
2        633  1002.604640
3        363   578.697900
4        319   518.798377

In [78]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [80]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [82]:
chequeator(submission)

You're ready to submit!
